Add Parquet IO library

michoecho opened the pull request:
Notes:
Most of the code was directly copied from Arrow, so it doesn't need much scrutiny. The relevant parts are situated in `namespace parquet::seastarized`, mainly in `column_reader.{h,cc}`, `column_writer.{h,cc}`, `file_reader.{h,cc}`, `file_writer.{h,cc}`, `io.{h,cc}`.
The old versions of classes in `parquet::seastarized` have been left in the code for now, so that we can compare them later with their seastarized versions. Ultimately they will be removed, together with `arrow/io/*`, `io_util{h.cc}` and `deprecated_io.{h,cc}`.
The giant `parquet_types.cc` and `parquet_types.h` are sources generated by Thrift. Eventually they should be generated during build.
For now, only Snappy compression is linked in. We will link in the rest (gzip, lzo, lz4, brotli, zstd) when we figure out the best way to add them to build.
The tests are written in GTest. We might rewrite them to Boost.Test to get rid of the dependency, someday.
All headers are in `include/` for now. We will sort out the public and private parts of the library after we have written something that uses it.

samvel1024 committed:
Port arrow/parquet to seastar (DOES NOT BUILD)

Moved column_writer.cc with possibly all dependencies.
Added compilation of snappy from sources.
Need to add thrift, openssl and seastarized tests.

samvel1024 committed:
Successfully compile libseastar, failing tests

samvel1024 committed:
Extract headers to /include dir, prepare the external interface for tests

samvel1024 committed:
Compile column writer test, TODO link gtest properly

samvel1024 committed:
Compile column writer test, fails with SEGV from ASAN

samvel1024 committed:
Fix column writer test, TODO Fix the mess in build definitions

samvel1024 committed:
Migrate parquet_file_serialize_test (Passing OK)

samvel1024 committed:
Migrate all tests, define a parquet_testing lib (Passing OK)

Commit:
Remove unnecessary parts of Arrow

samvel1024 committed:
Fix property ARROW_USE_OPENSSL

prawilny committed:
FutureInputStream private variables renamed

prawilny committed:
column_reader cleaned up and NextPage() fixed

prawilny committed:
unrelated typos

Commit:
Add some tests

Commit:
Minor indentation fixes

Commit:
Refactor NextPage and ReadBatch

Commit:
Fix broken indentation in column_scanner

Commit:
Fix c++14-incompatible changes

Commit:
Remove some unneeded code

Commit:
Delete UNION

Commit:
Remove builder_adaptive

psarna commented:
Thanks a lot! Few remarks:
1. I updated the zpp_parquet branch and rebased this series on top of it, so please use the updated version from now on. While rebasing, I also flattened the structure so it doesn't contain inner merges.
2. I currently cannot compile, because the linker can't find multiple symbols from libssl, and indeed I can't find traces of it being linked anywhere. Do you use a workaround for this? If so, please post it. it can also be a regular commit, as long as it makes the whole series compilable.

psarna commented:
hint: adding `OpenSSL::SSL` to cmake is enough

Commit:
Fix libcrypto linking

michoecho commented:
> I currently cannot compile, because the linker can't find multiple symbols from libssl

The linking order of `libseastar` and `libcrypto` was wrong. Fixed in f508b60

> I can't find traces of it being linked anywhere.

For some reason `libcrypto` was being linked directly to tests, not to seastar. Fixed in f508b60

I think it should work fine now.
The weird thing is: while 2d11ff1 failed for you, it didn't fail for me, even though the linking order was clearly wrong. It built cleanly on my machine (archlinux) and in fresh fedora:29 docker container. I don't understand g++'s behaviour there.
